File: go/v2/crypto/ecdh.go
package crypto

import (
	"crypto/ecdh"
	"crypto/ecdsa"
	"crypto/elliptic"
	"crypto/rand"
	"crypto/x509"
	"errors"
	"fmt"
	"math/big"
)

// ECDHComputeShared 计算 ECDH 共享秘密（P-256 X 坐标，32 字节）。
// privateKeyScalar: 32 字节 P-256 私钥标量（big-endian）
// peerPublicKeyDER: DER SubjectPublicKeyInfo 编码的对端公钥
//
// 输出为 P-256 共享点的 X 坐标 32 字节，与其它 SDK（Python/Node/C++）字节级一致。
func ECDHComputeShared(privateKeyScalar []byte, peerPublicKeyDER []byte) ([]byte, error) {
	if len(privateKeyScalar) != 32 {
		return nil, fmt.Errorf("ECDH: 私钥标量长度无效，期望 32 字节，实际 %d", len(privateKeyScalar))
	}

	curve := ecdh.P256()

	priv, err := curve.NewPrivateKey(privateKeyScalar)
	if err != nil {
		return nil, fmt.Errorf("ECDH: 解析私钥失败: %w", err)
	}

	peerPub, err := parseECDHPublicKeyDER(peerPublicKeyDER)
	if err != nil {
		return nil, fmt.Errorf("ECDH: 解析对端公钥失败: %w", err)
	}

	shared, err := priv.ECDH(peerPub)
	if err != nil {
		return nil, fmt.Errorf("ECDH: 计算共享秘密失败: %w", err)
	}

	if len(shared) != 32 {
		return nil, fmt.Errorf("ECDH: 共享秘密长度异常，期望 32 字节，实际 %d", len(shared))
	}

	return shared, nil
}

// GenerateP256Keypair 生成 P-256 密钥对。
// 返回 (privateKeyScalar 32 字节 big-endian, publicKeyDER SubjectPublicKeyInfo)。
func GenerateP256Keypair() ([]byte, []byte, error) {
	curve := ecdh.P256()
	priv, err := curve.GenerateKey(rand.Reader)
	if err != nil {
		return nil, nil, fmt.Errorf("ECDH: 生成密钥对失败: %w", err)
	}

	scalar := priv.Bytes()
	if len(scalar) != 32 {
		return nil, nil, fmt.Errorf("ECDH: 生成的私钥标量长度异常，期望 32，实际 %d", len(scalar))
	}

	pubDER, err := marshalECDHPublicKeyDER(priv.PublicKey())
	if err != nil {
		return nil, nil, fmt.Errorf("ECDH: 序列化公钥失败: %w", err)
	}

	return scalar, pubDER, nil
}

// PrivateToPublicDER 从私钥标量导出公钥 DER（SubjectPublicKeyInfo）。
func PrivateToPublicDER(privateKeyScalar []byte) ([]byte, error) {
	if len(privateKeyScalar) != 32 {
		return nil, fmt.Errorf("ECDH: 私钥标量长度无效，期望 32 字节，实际 %d", len(privateKeyScalar))
	}

	curve := ecdh.P256()
	priv, err := curve.NewPrivateKey(privateKeyScalar)
	if err != nil {
		return nil, fmt.Errorf("ECDH: 解析私钥失败: %w", err)
	}

	pubDER, err := marshalECDHPublicKeyDER(priv.PublicKey())
	if err != nil {
		return nil, fmt.Errorf("ECDH: 序列化公钥失败: %w", err)
	}
	return pubDER, nil
}

// parseECDHPublicKeyDER 解析 SubjectPublicKeyInfo DER 公钥并转为 *ecdh.PublicKey。
//
// x509.ParsePKIXPublicKey 对 EC 曲线返回 *ecdsa.PublicKey；必须经 .ECDH() 转为
// *ecdh.PublicKey 才能用于 crypto/ecdh 包的密钥协商。
func parseECDHPublicKeyDER(der []byte) (*ecdh.PublicKey, error) {
	pub, err := x509.ParsePKIXPublicKey(der)
	if err != nil {
		return nil, err
	}

	ecdsaPub, ok := pub.(*ecdsa.PublicKey)
	if !ok {
		return nil, errors.New("公钥类型不是 ECDSA P-256")
	}

	ecdhPub, err := ecdsaPub.ECDH()
	if err != nil {
		return nil, fmt.Errorf("ECDSA → ECDH 转换失败: %w", err)
	}
	return ecdhPub, nil
}

// marshalECDHPublicKeyDER 将 *ecdh.PublicKey 序列化为 SubjectPublicKeyInfo DER。
//
// x509.MarshalPKIXPublicKey 不直接接受 *ecdh.PublicKey，需先转为 *ecdsa.PublicKey。
// crypto/ecdh 的 PublicKey.Bytes() 返回 SEC1 未压缩点（0x04 || X || Y），可以从
// 中提取 X/Y 坐标构造 *ecdsa.PublicKey。
func marshalECDHPublicKeyDER(pub *ecdh.PublicKey) ([]byte, error) {
	raw := pub.Bytes()
	// SEC1 未压缩格式：0x04 || X(32) || Y(32) = 65 字节
	if len(raw) != 65 || raw[0] != 0x04 {
		return nil, fmt.Errorf("意外的 P-256 公钥编码：长度=%d, 首字节=0x%x", len(raw), raw[0])
	}

	ecdsaPub := &ecdsa.PublicKey{
		Curve: elliptic.P256(),
		X:     new(big.Int).SetBytes(raw[1:33]),
		Y:     new(big.Int).SetBytes(raw[33:65]),
	}

	return x509.MarshalPKIXPublicKey(ecdsaPub)
}

File: ts/src/v2/crypto/ecdh.ts
/**
 * ECDH P-256 — AUN E2EE V2 协议要求所有 SDK 的 ECDH 输出字节级一致。
 *
 * 共享秘密为 P-256 曲线点乘后的 X 坐标（32 字节，big-endian）。
 *
 * 实现选型：
 * - 使用 Node `crypto` 模块（非 WebCrypto，TS SDK 目标为 Node 环境）
 * - `createECDH('prime256v1')` 计算 X 坐标
 * - 公钥使用 DER SubjectPublicKeyInfo 编码（与 Python/Go SDK 对齐）
 * - 通过 JWK 中转完成 DER ↔ 未压缩点（0x04 || X || Y）的转换
 */

import { createECDH, createPublicKey } from 'node:crypto';

/**
 * 把 base64url 字符串解码为 32 字节大端序无前导零填充的字节数组。
 * 解决 JWK x/y 可能因高位为 0 而被截短的问题。
 */
function b64uToFixed32(b64url: string): Buffer {
  const buf = Buffer.from(b64url, 'base64url');
  if (buf.length === 32) return buf;
  if (buf.length < 32) {
    const padded = Buffer.alloc(32);
    buf.copy(padded, 32 - buf.length);
    return padded;
  }
  // 长度 > 32：去掉前导 0x00 填充
  const start = buf.length - 32;
  for (let i = 0; i < start; i++) {
    if (buf[i] !== 0) {
      throw new Error(`invalid EC coordinate: length=${buf.length}, leading non-zero byte`);
    }
  }
  return buf.subarray(start, buf.length);
}

/**
 * 把未压缩点（0x04 || X(32) || Y(32)）的 X/Y 坐标转换为 SPKI DER 公钥。
 */
function uncompressedPointToDer(uncompressed: Buffer): Buffer {
  if (uncompressed.length !== 65 || uncompressed[0] !== 0x04) {
    throw new Error('invalid uncompressed P-256 point');
  }
  const x = uncompressed.subarray(1, 33);
  const y = uncompressed.subarray(33, 65);
  const pubKey = createPublicKey({
    key: {
      kty: 'EC',
      crv: 'P-256',
      x: x.toString('base64url'),
      y: y.toString('base64url'),
    },
    format: 'jwk',
  });
  return pubKey.export({ format: 'der', type: 'spki' });
}

/**
 * 计算 ECDH 共享秘密（P-256 X 坐标，32 字节）。
 *
 * @param privateKeyScalar 32 字节 P-256 私钥标量（big-endian）
 * @param peerPublicKeyDer DER SubjectPublicKeyInfo 编码的对端公钥
 * @returns 32 字节共享秘密（X 坐标 big-endian）
 */
export function ecdhComputeShared(
  privateKeyScalar: Uint8Array,
  peerPublicKeyDer: Uint8Array,
): Uint8Array {
  if (privateKeyScalar.length !== 32) {
    throw new Error(`expected 32-byte P-256 private scalar, got ${privateKeyScalar.length}`);
  }

  const ecdh = createECDH('prime256v1');
  ecdh.setPrivateKey(Buffer.from(privateKeyScalar));

  // 解析 DER SPKI 公钥 → 提取未压缩点（0x04 || X || Y）
  const peerPubKey = createPublicKey({
    key: Buffer.from(peerPublicKeyDer),
    format: 'der',
    type: 'spki',
  });
  const jwk = peerPubKey.export({ format: 'jwk' });
  if (jwk.kty !== 'EC' || jwk.crv !== 'P-256' || !jwk.x || !jwk.y) {
    throw new Error('peer public key is not a P-256 EC key');
  }
  const x = b64uToFixed32(jwk.x);
  const y = b64uToFixed32(jwk.y);
  const uncompressed = Buffer.concat([Buffer.from([0x04]), x, y]);

  // ECDH.computeSecret 默认返回 X 坐标（32 字节大端序）
  const shared = ecdh.computeSecret(uncompressed);
  return new Uint8Array(shared);
}

/**
 * 生成 P-256 密钥对。
 *
 * @returns [privateKeyScalar 32B, publicKeyDer SPKI]
 */
export function generateP256Keypair(): [Uint8Array, Uint8Array] {
  const ecdh = createECDH('prime256v1');
  const pubUncompressed = ecdh.generateKeys();
  const priv = ecdh.getPrivateKey();

  // setPrivateKey/generateKeys 返回的 priv 可能少于 32 字节（高位为 0），左零填充
  let privPadded: Buffer;
  if (priv.length === 32) {
    privPadded = priv;
  } else if (priv.length < 32) {
    privPadded = Buffer.alloc(32);
    priv.copy(privPadded, 32 - priv.length);
  } else {
    throw new Error(`unexpected P-256 private key length: ${priv.length}`);
  }

  const pubDer = uncompressedPointToDer(pubUncompressed);
  return [new Uint8Array(privPadded), new Uint8Array(pubDer)];
}

/**
 * 从私钥标量导出公钥 DER（SPKI 编码）。
 */
export function privateToPublicDer(privateKeyScalar: Uint8Array): Uint8Array {
  if (privateKeyScalar.length !== 32) {
    throw new Error(`expected 32-byte P-256 private scalar, got ${privateKeyScalar.length}`);
  }
  const ecdh = createECDH('prime256v1');
  ecdh.setPrivateKey(Buffer.from(privateKeyScalar));
  const pubUncompressed = ecdh.getPublicKey();
  const pubDer = uncompressedPointToDer(pubUncompressed);
  return new Uint8Array(pubDer);
}
